src/registers/debug.rs

```diff
@@ -0,0 +1,580 @@
+//! Functions to read and write debug registers.
+
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use core::convert::TryFrom;
+use core::ops::Range;
+
+use crate::VirtAddr;
+
+use bit_field::BitField;
+use bitflags::bitflags;
+
+/// Debug Address Register
+///
+/// Holds the address of a hardware breakpoint.
+pub trait DebugAddressRegister {
+    /// Reads the current breakpoint address.
+    fn read() -> VirtAddr;
+
+    /// Writes the provided breakpoint address.
+    fn write(addr: VirtAddr);
+}
+
+macro_rules! debug_address_register_struct {
+    ($Dr:ident) => {
+        /// Debug Address Register
+        ///
+        /// Holds the address of a hardware breakpoint.
+        #[derive(Debug)]
+        pub struct $Dr;
+    };
+}
+
+debug_address_register_struct!(Dr0);
+debug_address_register_struct!(Dr1);
+debug_address_register_struct!(Dr2);
+debug_address_register_struct!(Dr3);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+/// A valid debug address register number.
+///
+/// Must be between 0 and 3 (inclusive).
+pub struct DebugAddressRegisterNumber(u8);
+
+impl DebugAddressRegisterNumber {
+    /// Creates a debug address register number without checking the value.
+    ///
+    /// # Safety
+    ///
+    /// The value must be between 0 and 3 (inclusive).
+    pub const unsafe fn new_unchecked(n: u8) -> Self {
+        Self(n)
+    }
+
+    /// Creates a debug address register number if it is valid.
+    pub const fn new(n: u8) -> Option<Self> {
+        match n {
+            0..=3 => Some(Self(n)),
+            _ => None,
+        }
+    }
+
+    /// Returns the number as a primitive type.
+    pub const fn get(self) -> u8 {
+        self.0
+    }
+}
+
+/// The error type returned when a checked integral type conversion fails.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct TryFromIntError(());
+
+impl From<core::num::TryFromIntError> for TryFromIntError {
+    fn from(_: core::num::TryFromIntError) -> Self {
+        Self(())
+    }
+}
+
+impl TryFrom<u8> for DebugAddressRegisterNumber {
+    type Error = TryFromIntError;
+
+    fn try_from(n: u8) -> Result<Self, Self::Error> {
+        Self::new(n).ok_or(TryFromIntError(()))
+    }
+}
```

Comment:
Does this `TryFrom` implementation add value we don't get with the inherent constructor functions? We currently don't  provide `TryFrom` implementations for other similar types e.g. `PageTableIndex`, `PageOffset`, `PageTableLevel` and `PrivilegeLevel`.

Reply:
No, I removed it. :+1: